permitchecklog.go

```diff
@@ -0,0 +1,204 @@
+	config           ConfigUsageTracking
+	log              *log.Logger
+	logs             []CheckLogEntry
+	mutex            sync.RWMutex
+	stopChan         chan struct{}
+	doneChan         chan struct{}
+	flushDoneChan    chan struct{}
+	stopOnce         sync.Once
+	flushTicker      *time.Ticker
+	flushing         bool // Track if a flush is in progress
+	db               *sql.DB
+	droppedLogsCount int64
+	memDump          []string // For testing purposes
+}
+
+// NewCheckUsageTracker creates a new usage tracker instance
+func NewCheckUsageTracker(c ConfigUsageTracking, log *log.Logger, db *sql.DB) *CheckUsageTracker {
+	c.SetDefaults()
+	tracker := &CheckUsageTracker{
+		config:        c,
+		log:           log,
+		logs:          make([]CheckLogEntry, 0),
+		stopChan:      make(chan struct{}),
+		doneChan:      make(chan struct{}),
+		flushDoneChan: make(chan struct{}, 1),
+		db:            db, // Database connection will be set later
+	}
+	if tracker.config.Test_MemDump {
+		tracker.memDump = make([]string, 0)
+	}
+	return tracker
+}
+
+func (t *CheckUsageTracker) Initialize(logger *log.Logger) {
+	if logger != nil {
+		t.log = logger
+
+		if t.config.Enabled {
+			t.start()
+		}
+
+		if t.db == nil {
+			t.log.Warn("CheckUsageTracker initialized without a database connection, logs will not be persisted")
+		} else {
+			t.log.Info("CheckUsageTracker initialized with database connection")
+		}
+	}
+}
+
+// LogCheck adds a check request to the in-memory log
+func (t *CheckUsageTracker) LogCheck(sessionToken string, token *Token) {
+	if !t.config.Enabled {
+		return
+	}
+
+	entry := CheckLogEntry{
+		Timestamp:    time.Now().UTC(),
+		SessionToken: sessionToken,
+		UserId:       token.UserId,
+	}
+
+	t.mutex.Lock()
+	if len(t.logs) >= t.config.MaxEntries {
+		t.droppedLogsCount++
+		// drop entry
+	} else {
+		t.logs = append(t.logs, entry)
+	}
+	t.mutex.Unlock()
+}
+
+// start begins the periodic flush process
+func (t *CheckUsageTracker) start() {
+	flushInterval := time.Duration(t.config.FlushIntervalSeconds) * time.Second
+	if flushInterval <= 0 {
+		flushInterval = 60 * time.Second // Default to 1 minute
+	}
+
+	t.flushTicker = time.NewTicker(flushInterval)
+
+	go func() {
+		defer close(t.doneChan)
+		for {
+			select {
+			case <-t.flushTicker.C:
+				t.flush()
+			case <-t.stopChan:
+				t.flushTicker.Stop()
+				t.flush() // Final flush before stopping
+				return
+			}
+		}
+	}()
+}
+
+// Stop gracefully shuts down the usage tracker
+func (t *CheckUsageTracker) Stop() {
+	t.stopOnce.Do(func() {
+		if t.flushTicker != nil && t.stopChan != nil {
+			close(t.stopChan)
+			<-t.doneChan      // Wait for main flush management goroutine to finish
+			<-t.flushDoneChan // Wait for any ongoing flush operation to complete
+		}
+	})
+}
+
+// flush writes the in-memory logs to persistent storage and clears the memory
+// TODO : Implement a more generic persistence abstraction (e.g. Provider pattern)
+func (t *CheckUsageTracker) flush() {
```

Comment:
Bad bot. This entire function is the result of using an array instead of a channel to manage the `*CheckLogEntry`s. Please remove all of this unnecessary complexity (delete the function).